dali_tf_plugin/dali_dataset_op.cc

```diff
@@ -284,6 +323,111 @@ class DALIDatasetOp : public DatasetOpKernel {
       }
 
      private:
+
+      Status GetCompatibleShape(TensorShape &result, const PartialTensorShape &required_shape,
+                                const TensorShape &dali_shape, int batch_size, int output_idx) {
+        if (required_shape.IsCompatibleWith(dali_shape)) {
+          result = dali_shape;
+          return Status::OK();
+        }
+
+        // both ranks should be known at this points (otherwise shapes are compatible)
+        // If they are not compatible and have the same rank or the required shape is the bigger one
+        // we cannot make them compatible.
+        if (required_shape.dims() >= dali_shape.dims()) {
+          std::stringstream ss;
+          ss << "The shape provided for output `" << output_idx << "` is not compatible with "
+              << "the shape returned by DALI Pipeline. Expected (output_shapes[" << output_idx
+              << "]): " << required_shape << ", got from Pipeline: " << dali_shape << ".";
+          return errors::InvalidArgument(ss.str());
+        }
+        for (int i = 0; i < required_shape.dims(); i++) {
+          result.AddDim(0);
+        }
+        // Non-trivial batch size case, diffrent dims
+        if (batch_size != 1) {
+          // Should not happen, batch size from DALI will always match
+          if (dali_shape.dim_size(0) != batch_size) {
+            std::stringstream ss;
+            ss << "The shape returned by DALI Pipeline for output `" << output_idx
+                << "` has different `batch_size` than the one specified in `DALIDataset`. "
+                << "Specified `batch_size`: " << batch_size
+                << ", got from Pipeline: " << dali_shape.dim_size(0) << " in shape: "
+                << dali_shape << ".";
+            return errors::InvalidArgument(ss.str());
+          }
+          // It's easier to check it in C++ as in Python the structure is a bit convoluted
+          if (!DimSizeMatch(required_shape.dim_size(0), batch_size)) {
+            std::stringstream ss;
+            ss << "The shape provided for output `" << output_idx << "` is not compatible with "
+                << "the `batch_size` argument that was specified in `DALIDataset`. "
+                << "Specified `batch_size`: " << batch_size << ", got: "
+                << required_shape.dim_size(0) << " in shape: " << required_shape << ".";
+            return errors::InvalidArgument(ss.str());
+          }
+        }
+        // Special case for 1-element tensors
+        if (dali_shape.num_elements() == 1) {
+          TensorShape regular_shape;
+          if (required_shape.AsTensorShape(&regular_shape) && regular_shape.num_elements() == 1) {
+            result = regular_shape;
+            return Status::OK();
+          }
+        }
+        int matches = CountShapeMatches(result, required_shape, dali_shape);
+        if (matches != 1) {
```

Comment:
```suggestion
        if (matches > 1) {
```
don't you mean that?

Reply:
When there is 0 matches we still don't have a valid match.